### PR TITLE
feat: automated generate icons script

### PR DESCRIPTION
**Pull request by georgewrmarshall**
## **Description**

This PR adds an automated icon generation system that converts SVG assets into React components using [SVGR](https://react-svgr.com/docs). Key improvements include:

1. Reason for change: Currently, SVG icons need to be manually converted into React components, which is time-consuming and error-prone.
2. Solution: Automated pipeline that:
   - Converts SVG files to typed React components using SVGR
   - Generates an index file for easy importing
   - Updates `IconName` enum automatically
   - Maintains consistent component structure and props

## **Related issues**

Part of: https://github.com/MetaMask/metamask-design-system/issues/182

## **Manual testing steps**

1. Add a new SVG file to `packages/design-system-react/src/components/icon/assets/`
2. Run `yarn generate:icons` in the design-system-react package
3. Verify that:
   - A new React component is generated in the icons/ directory
   - The component is exported in the index.ts file
   - The IconName enum is updated with the new icon
   - The icon can be used in the Icon component

### **Before**
Manual SVG implementation directly in Icon.tsx:

https://github.com/user-attachments/assets/96e5b382-44a4-43a3-bb82-80e688d9fdbc

### **After**
Automated component generation and usage:

https://github.com/user-attachments/assets/10cefb5d-6a95-4488-a263-ff69de856533

https://github.com/user-attachments/assets/cc185de6-4245-4cc3-85d0-7bfe5a719028

## **Pre-merge author checklist**

- [x] I've followed MetaMask Contributor Docs
- [x] I've completed the PR template to the best of my ability
- [x] I've included tests if applicable
- [x] I've documented my code using JSDoc format if applicable
- [x] I've applied the right labels on the PR

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: packages/design-system-react/src/components/text/README.mdx [hunk 1]
```diff
@@ -6,6 +6,12 @@ import * as TextStories from './Text.stories';
```
Comment (georgewrmarshall) [inline packages/design-system-react/src/components/text/README.mdx:6]: Move this to a separate PR

File: packages/design-system-react/src/components/text/Text.stories.tsx [hunk 2]
```diff
@@ -21,6 +21,11 @@ const meta: Meta<typeof Text> = {
       page: README,
```
Comment (georgewrmarshall) [inline packages/design-system-react/src/components/text/Text.stories.tsx:21]: Move this to a separate PR

File: .gitignore [hunk 3]
```diff
@@ -60,3 +60,6 @@ yarn-error.*
 
 # local env files
 .env*.local
+
+# Auto-generated icons
+packages/design-system-react/src/components/icon/icons/
```
Comment (georgewrmarshall) [inline .gitignore:65]: These icon files are automatically generated by SVGR during the build process and don't need to be tracked in git.

File: packages/design-system-react/.eslintrc.js [hunk 4]
```diff
@@ -16,4 +16,5 @@ module.exports = {
       },
     },
   ],
+  ignorePatterns: ['src/components/icon/icons/*'], // Auto-generated icons
```
Comment (georgewrmarshall) [inline packages/design-system-react/.eslintrc.js:19]: Adding `src/components/icon/icons/*` to `ignorePatterns` because SVGR's generation process can reorder imports differently than our lint rules expect. Since these are auto-generated files, we don't need to enforce our lint rules on them.

File: packages/design-system-react/jest.config.js [hunk 5]
```diff
@@ -18,7 +18,8 @@ module.exports = merge(baseConfig, {
   coveragePathIgnorePatterns: [
     'index.ts',
     '.d.ts',
-    'scripts/create-component/ComponentName/',
+    'src/components/icon/icons/', // Auto-generated icons
```
Comment (georgewrmarshall) [inline packages/design-system-react/jest.config.js:21]: Ignoring autogenerated icons from unit tests

File: packages/design-system-react/package.json [hunk 6]
```diff
@@ -35,10 +35,13 @@
     "dist/"
   ],
   "scripts": {
-    "build": "ts-bridge --project tsconfig.build.json --verbose --clean --no-references",
+    "build": "yarn generate:icons && ts-bridge --project tsconfig.build.json --verbose --clean --no-references",
```
Comment (georgewrmarshall) [inline packages/design-system-react/package.json:38]: Adding `generate:icons` to the build script to ensure SVG icons are automatically converted to React components before the TypeScript build starts. This guarantees that all icon components are available during the build process and prevents missing component errors.

File: packages/design-system-react/package.json [hunk 7]
```diff
@@ -35,10 +35,13 @@
...
     "changelog:update": "../../scripts/update-changelog.sh @metamask/design-system-react",
     "changelog:validate": "../../scripts/validate-changelog.sh @metamask/design-system-react",
+    "clean:icons": "rimraf src/components/icon/icons",
     "create-component": "ts-node scripts/create-component",
+    "generate:icons": "yarn clean:icons && svgr --config-file src/components/icon/.svgrrc.js -d src/components/icon/icons src/components/icon/assets/*.svg && yarn generate:icons:index",
+    "generate:icons:index": "ts-node scripts/generate-icons-index.ts",
```
Comment (georgewrmarshall) [inline packages/design-system-react/package.json:44]: These scripts manage our icon generation pipeline:
- clean:icons: Removes previously generated icon components to ensure a clean slate
- generate:icons: Main icon generation script that:
1. Cleans existing icons
2. Uses SVGR to convert SVG files from assets/ into React components
3. Generates an index file for easy importing
- generate:icons:index: Creates an index file that exports all generated icon components for convenient access

File: packages/design-system-react/package.json [hunk 8]
```diff
@@ -51,10 +54,12 @@
     "tailwind-merge": "^2.0.0"
   },
   "devDependencies": {
+    "@jest/globals": "^29.7.0",
```
Comment (georgewrmarshall) [inline packages/design-system-react/package.json:57]: Used for generite-icons-index tests

File: packages/design-system-react/package.json [hunk 9]
```diff
@@ -65,6 +70,7 @@
     "deepmerge": "^4.2.2",
     "jest": "^29.7.0",
     "jest-environment-jsdom": "^29.7.0",
+    "rimraf": "^5.0.5",
```
Comment (georgewrmarshall) [inline packages/design-system-react/package.json:73]: removes the icons/ folder

File: packages/design-system-react/scripts/generate-icons-index.test.ts [hunk 10]
```diff
@@ -0,0 +1,145 @@
+import { jest } from '@jest/globals';
```
Comment (georgewrmarshall) [inline packages/design-system-react/scripts/generate-icons-index.test.ts:1]: Test for the generate-icons-index file

File: packages/design-system-react/scripts/generate-icons-index.ts [hunk 11]
```diff
@@ -0,0 +1,86 @@
+import * as fs from 'fs/promises';
```
Comment (georgewrmarshall) [inline packages/design-system-react/scripts/generate-icons-index.ts:1]: Automatically generates the `index.ts` file in the icons/

File: packages/design-system-react/src/components/icon/.svgrrc.js [hunk 12]
```diff
@@ -0,0 +1,28 @@
+module.exports = {
```
Comment (georgewrmarshall) [inline packages/design-system-react/src/components/icon/.svgrrc.js:1]: This configuration file controls how SVGR transforms our SVG files into React components. It:
- Enables TypeScript output
- Adds React ref forwarding for component composition
- Sets consistent SVG props (role="img" and fill="currentColor")
- Uses our custom template for consistent component structure
- Configures SVGO optimization settings to preserve viewBox and ensure consistent color handling

This ensures all our icon components are generated with consistent typing, accessibility, and styling capabilities.

File: packages/design-system-react/src/components/icon/Icon.test.tsx [hunk 13]
```diff
@@ -88,10 +88,17 @@ describe('Icon error cases', () => {
...
     expect(container.firstChild).toBeNull();
   });
+
+  it('should warn and return null when icon is not found', () => {
+    const { container } = render(<Icon name={'NonExistentIcon' as IconName} />);
+
+    expect(consoleSpy).toHaveBeenCalledWith('Icon "NonExistentIcon" not found');
+    expect(container.firstChild).toBeNull();
+  });
 });
```
Comment (georgewrmarshall) [inline packages/design-system-react/src/components/icon/Icon.test.tsx:104]: Adding error handling tests for the Icon component:
1. Tests the case when name prop is missing:
- Uses @ts-expect-error to intentionally pass undefined name
- Verifies warning message is shown
- Confirms component returns null

2. Tests the case when an invalid icon name is provided:
- Attempts to render non-existent icon
- Verifies appropriate warning message
- Confirms component returns null

These tests ensure our Icon component fails gracefully with helpful warning messages instead of throwing errors.

File: packages/design-system-react/src/components/icon/Icon.tsx [hunk 14]
```diff
@@ -25,15 +34,10 @@ export const Icon: React.FC<IconProps> = ({
...
       className={mergedClassName}
       {...(props as React.SVGProps<SVGSVGElement>)}
       style={style}
-      xmlns="http://www.w3.org/2000/svg"
-      viewBox="0 0 512 512"
-      fill="currentColor"
-    >
-      <path d="m337 51l-171 0c-75 0-119 45-119 119l0 172c0 74 44 119 119 119l171 0c75 0 119-45 119-119l0-172c1-74-44-119-119-119z m-3 220l-67 0 0 67c0 8-7 15-15 15-9 0-16-7-16-15l0-67-66 0c-9 0-16-7-16-15 0-8 7-15 16-15l66 0 0-67c0-8 7-15 16-15 8 0 15 7 15 15l0 67 67 0c8 0 15 7 15 15 0 8-7 15-15 15z" />
-    </svg>
+    />
```
Comment (georgewrmarshall) [inline packages/design-system-react/src/components/icon/Icon.tsx:41]: Using dynamic component lookup (Icons[name]) instead of static SVG imports:
- Allows lazy loading of icons - only the icons that are actually used are included in the bundle
- Provides type safety through the IconName enum
- Makes it easy to add new icons without modifying the Icon component code
- Centralizes icon rendering logic and props handling in one place

typescript> const IconComponent = Icons[name];> if (!IconComponent) {> console.warn(Icon "${name}" not found);> return null;> }>
This pattern ensures consistent rendering and proper error handling while keeping our bundle size optimized.

File: packages/design-system-react/src/components/icon/assets/add-square.svg
Comment (georgewrmarshall) [inline packages/design-system-react/src/components/icon/assets/add-square.svg:1]: Moving `add-square.svg` assets into assets directory. More svg icons assets will be added in a subsequent PR

File: packages/design-system-react/src/components/icon/template.d.ts [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+  tpl: (strings: TemplateStringsArray, ...values: unknown[]) => string;
+  options?: Record<string, unknown>;
+};
+
+declare const template: (
+  variables: TemplateVariables,
+  context: TemplateContext,
+) => string;
+
+export default template;
```
Comment (georgewrmarshall) [inline packages/design-system-react/src/components/icon/template.d.ts:20]: The template.d.ts file is needed alongside template.js because:
1. SVGR requires the template file to be JavaScript (.js), as it's used in their build pipeline
2. However, we still want TypeScript type checking and IDE support when working with the template
3. The .d.ts declaration file provides type definitions for:
- Template variables (imports, interfaces, component name, etc.)
- Template context (tpl function and options)
This setup gives us the best of both worlds: compatibility with SVGR's requirements while maintaining type safety in our codebase.

File: packages/design-system-react/src/components/icon/template.js [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+    ${variables.jsx}
+  </svg>
+);
+
+const ForwardRef = forwardRef(${variables.componentName});
+export default ForwardRef;
+`;
+}
+
+module.exports = template;
```
Comment (georgewrmarshall) [inline packages/design-system-react/src/components/icon/template.js:53]: The template.js file is a crucial part of our SVGR pipeline:
1. It's a custom template that controls how SVG files are transformed into React components
2. Must be JavaScript because SVGR's build process directly require()s and executes this file during transformation
3. The template:
- Sets up proper TypeScript types for the generated components
- Adds consistent props handling (SVGProps, ref forwarding)
- Ensures consistent SVG attributes (viewBox, xmlns, etc.)
- Maintains uniform component structure across all icons

File: packages/design-system-react/src/components/text/README.mdx (hunk 1 above)
Comment (georgewrmarshall) [reply packages/design-system-react/src/components/text/README.mdx:6]: Moved here https://github.com/MetaMask/metamask-design-system/pull/276

File: packages/design-system-react/src/components/text/Text.stories.tsx (hunk 2 above)
Comment (georgewrmarshall) [reply packages/design-system-react/src/components/text/Text.stories.tsx:21]: Moved here https://github.com/MetaMask/metamask-design-system/pull/276

File: yarn.config.cjs [hunk 17]
```diff
@@ -95,20 +95,6 @@ module.exports = defineConfig({
...
-              'scripts.build',
-              'ts-bridge --project tsconfig.build.json --verbose --clean --no-references',
-            );
-          } else if (workspace.ident === '@metamask/design-tokens') {
-            expectWorkspaceField(
-              workspace,
-              'scripts.build',
-              'ts-bridge --project tsconfig.build.json --verbose --clean --no-references && yarn build:css',
-            );
-          }
```
Comment (georgewrmarshall) [inline yarn.config.cjs:111]: Removing standardized build script checks from constraints because:
1. Our packages have diverse build requirements:
- Design tokens needs CSS generation
- React components need SVGR icon generation
- React Native has its own build process
2. Forcing a single build command pattern (ts-bridge) across all packages doesn't make sense when build steps vary significantly
3. Each package should be able to define its own build process based on its specific needs

This change allows packages to maintain build scripts that are appropriate for their individual requirements rather than forcing a one-size-fits-all approach.

File: packages/design-system-react/src/components/icon/icons/AddSquare.tsx [hunk 18]
```diff
@@ -0,0 +1,20 @@
+import * as React from 'react';
```
Comment (georgewrmarshall) [inline packages/design-system-react/src/components/icon/icons/AddSquare.tsx:1]: In a previous version of this PR I added the `icons/` folder to the .gitignore because they are all automatically generated but we need them for tests

File: packages/design-system-react/src/components/icon/icons/index.ts [hunk 19]
```diff
@@ -0,0 +1,14 @@
+// This file is auto-generated. Do not edit manually
```
Comment (georgewrmarshall) [inline packages/design-system-react/src/components/icon/icons/index.ts:1]: The `index` file generated by the `generate-icons-index` script was originally ignored in Git, but we need it to be included for tests.

**Comment by georgewrmarshall** [comment]
@SocketSecurity ignore npm/@svgr/cli@8.1.0
- Protected against supply chain attacks by lavamoat

File: packages/design-system-react/src/components/icon/Icon.stories.tsx [hunk 20]
```diff
@@ -81,7 +81,11 @@ export const Default: Story = {
               key={iconKey}
               className="flex flex-col items-center justify-center p-4 border rounded"
             >
-              <Icon className="mb-2" {...args} />
+              <Icon
+                className="mb-2"
+                {...args}
+                name={IconName[iconKey as keyof typeof IconName]}
```
Comment (georgewrmarshall) [inline packages/design-system-react/src/components/icon/Icon.stories.tsx:87]: Updates `Icon` to use different icons via IconName

**Comment by georgewrmarshall** [comment]
@SocketSecurity ignore-all
- Protected against supply chain attacks by lavamoat

Screenshot 2024-12-11 at 1 08 32 PM

**Review by brianacnguyen — APPROVED** [review]
